plugins/events/natsjs/nats.go

```diff
@@ -0,0 +1,233 @@
+	}
+	for _, o := range opts {
+		o(&options)
+	}
+
+	s := &stream{opts: options}
+	natsJetStreamCtx, err := connectToNatsJetStream(options)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to nats cluster %v: %v", options.ClusterID, err)
+	}
+	s.natsJetStreamCtx = natsJetStreamCtx
+	return s, nil
+}
+
+type stream struct {
+	opts             Options
+	natsJetStreamCtx nats.JetStreamContext
+}
+
+func connectToNatsJetStream(options Options) (nats.JetStreamContext, error) {
+	nopts := nats.GetDefaultOptions()
+	if options.TLSConfig != nil {
+		nopts.Secure = true
+		nopts.TLSConfig = options.TLSConfig
+	}
+	if len(options.Address) > 0 {
+		nopts.Servers = strings.Split(options.Address, ",")
+	}
+
+	conn, err := nopts.Connect()
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to nats at %v with tls enabled (%v): %v", options.Address, nopts.TLSConfig != nil, err)
+	}
+
+	js, err := conn.JetStream()
+	if err != nil {
+		return nil, fmt.Errorf("error while obtaining JetStream context: %v", err)
+	}
+
+	return js, nil
+}
+
+// Publish a message to a topic
+func (s *stream) Publish(topic string, msg interface{}, opts ...events.PublishOption) error {
+	// validate the topic
+	if len(topic) == 0 {
+		return events.ErrMissingTopic
+	}
+
+	// parse the options
+	options := events.PublishOptions{
+		Timestamp: time.Now(),
+	}
+	for _, o := range opts {
+		o(&options)
+	}
+
+	// encode the message if it's not already encoded
+	var payload []byte
+	if p, ok := msg.([]byte); ok {
+		payload = p
+	} else {
+		p, err := json.Marshal(msg)
+		if err != nil {
+			return events.ErrEncodingMessage
+		}
+		payload = p
+	}
+
+	// construct the event
+	event := &events.Event{
+		ID:        uuid.New().String(),
+		Topic:     topic,
+		Timestamp: options.Timestamp,
+		Metadata:  options.Metadata,
+		Payload:   payload,
+	}
+
+	// serialize the event to bytes
+	bytes, err := json.Marshal(event)
+	if err != nil {
+		return errors.Wrap(err, "Error encoding event")
+	}
+
+	pubOpts := []nats.PubOpt{
+		// TODO: to make de-duplication work, we need to pass the event from the outside as an option
+		// nats.MsgId(event.ID), // event de-duplication
+	}
+
+	// publish the event to the topic's channel
+	if _, err := s.natsJetStreamCtx.PublishAsync(event.Topic, bytes, pubOpts...); err != nil {
+		return errors.Wrap(err, "Error publishing message to topic")
+	}
+
+	return nil
+}
+
+// Consume from a topic
+func (s *stream) Consume(topic string, opts ...events.ConsumeOption) (<-chan events.Event, error) {
+	// validate the topic
+	if len(topic) == 0 {
+		return nil, events.ErrMissingTopic
+	}
+
+	// parse the options
+	options := events.ConsumeOptions{
+		Group: uuid.New().String(),
+	}
+	for _, o := range opts {
+		o(&options)
+	}
+
+	// setup the subscriber
+	c := make(chan events.Event)
+	handleMsg := func(m *nats.Msg) {
+
+		ctx, cancel := context.WithCancel(context.TODO())
+		defer cancel()
+
+		// TODO: not supported by go-micro interface
```

Comment:
@asim NATS Jetstream supports "stretching" the message ACK timeout (`options.AckWait`) but that would need exposing a `InProgressFunc` to the events channel consumer. Is that something we want and can support with other implementations? see https://docs.nats.io/using-nats/developer/develop_jetstream/model_deep_dive -> `AckProgress`

Reply:
I think its best to do it in an isolated case specifically for this. You can see the rabbitmq broker plugin for implementation specific options that can be passed through. It's not clear that other implementations would support it so we'd inherit an option nothing can deal with.